Update fixtures.rst w/ finalizer order (#10171)
pytest-dev/pytest: doc/en/how-to/fixtures.rst

It's a bit longer than yield fixtures and a bit more complex, but it
does offer some nuances for when you're in a pinch.

Note on finalizer order
""""""""""""""""""""""""

Finalizers are executed in a first-in-last-out order, while operations after yield are executed sequentially.
Consider the differences in the following examples:

> 

Reviewer: This will cleanup the files from the previous example.

```suggestion

.. regendoc:wipe

```

.. code-block:: python

    import pytest
    from functools import partial

    @pytest.fixture
    def fix_w_finalizers(request):
        request.addfinalizer(partial(print, "finalizer_2"))
        request.addfinalizer(partial(print, "finalizer_1"))

    @pytest.fixture
    def fix_w_yield():
        yield

>         print("after_yield_1")

Reviewer: This is not a good example I think, because of course both prints will be executed sequentially (pytest cannot mess with how Python executes code within a function).

I suggest we use two fixtures:

```python
    @pytest.fixture
    def fix_w_yield1():
        yield
        print("after_yield_1")

    @pytest.fixture
    def fix_w_yield2():
        yield
        print("after_yield_2")
```

And then make `test_bar` request both:

```python
    def test_bar(fix_w_yield1, fix_w_yield2):
        print("test_bar")
```

Author: Ah, I see.
The execution order with the addfinalizer method is evident, so the execution order only needs clarification in this case.
Lovely! How about now?

Reviewer: I think we can keep the examples using `addfinalizer` as they are useful for comparision, so could you please bring them back? Thanks

Author: Sure thing, re-added.

        print("after_yield_2")

    def test_foo(fix_w_finalizers):
        print("test_foo")

>         pass

Reviewer: We don't need the `pass` statement.

```suggestion
```

    def test_bar(fix_w_yield):
        print("test_bar")

>         pass

Reviewer: ```suggestion
```

.. code-block:: pytest

    $ pytest test_module.py
    =========================== test session starts ============================
    platform linux -- Python 3.x.y, pytest-7.x.y, pluggy-1.x.y
    collected 2 items

    main.py
    test_foo
    .finalizer_1
    finalizer_2

    test_bar
    .after_yield_1
    after_yield_2

> .. code-block:: pytest

Reviewer: This block should be moved above the `Note on finalizer order` I believe?

Author: Ah, yes. Moved!

.. code-block:: pytest

   $ pytest -q test_emaillib.py
  .                                                                    [100%]
  1 passed in 0.12s

Note on finalizer order
""""""""""""""""""""""""

> Finalizers are executed in a first-in-last-out order, while operations after yield are executed sequentially.

Reviewer: >  while operations after yield are executed sequentially.

This sentence needs tweaking if we look at the output. 👍 

Finalizers are executed in FILO order, that is correct, because we can see that adding `fin2` and `fin1` we get back `fn1`, `fn2` in the output.

But the same can be said for the yield fixtures: we request them in `fix_w_yield1`, `fix_w_yield2` order, and we get `fix_w_yield2` and `fix_w_yield1` in the output.

That's because `yield` fixtures use `addfinalizer` behind the scenes: when the fixture executes, we register a function with `addfinalizer` that resumes the generator (which in turn calls the teardown code).

Not sure how to "fix" this sentence, or perhaps we can explain something along the lines of what I wrote above, if that makes things clearer.

Author: Yeah, that makes sense.
I tried to explain how both examples show the similar FILO behavior, plus added the hidden usage of addfinalizer after the examples.
What do you think?

.. code-block:: pytest

   $ pytest -q test_emaillib.py
  .                                                                    [100%]
  1 passed in 0.12s

Note on finalizer order
""""""""""""""""""""""""

Finalizers are executed in a first-in-last-out order.

> For yield fixtures, the first fixture to run is the right-most one, i.e. the last test parameter.

Reviewer: ```suggestion
For yield fixtures, the first teardown code to run is from the right-most fixture, i.e. the last test parameter.
```

The fixture setup code is executed left-to-right, while the teardown code is right-to-left.

Author: That's right, thanks for clarifying :)
